hydra-node/test/Hydra/API/ServerSpec.hs

```diff
@@ -99,21 +107,87 @@ spec = parallel $ do
+            withClient port "/" $ \conn -> do
               received <- replicateM (length outputs + 1) (receiveData conn)
               case traverse Aeson.eitherDecode received of
                 Left{} -> failure $ "Failed to decode messages:\n" <> show received
-                Right timedOutputs ->
+                Right timedOutputs -> do
                   (output <$> timedOutputs) `shouldBe` greeting : outputs
 
+  it "does not echo history if client says no" $
+    checkCoverage . monadicIO $ do
+      history :: [ServerOutput SimpleTx] <- pick arbitrary
+      monitor $ cover 0.1 (null history) "no message when reconnecting"
+      monitor $ cover 0.1 (length history == 1) "only one message when reconnecting"
+      monitor $ cover 1 (length history > 1) "more than one message when reconnecting"
+      run $ do
+        withFreePort $ \port ->
+          withAPIServer @SimpleTx "127.0.0.1" (fromIntegral port) alice mockPersistence nullTracer noop $ \Server{sendOutput} -> do
+            let sendFromApiServer = sendOutput
+            mapM_ sendFromApiServer history
+            -- start client that doesn't want to see the history
+            withClient port "/?history=no" $ \conn -> do
+              -- wait on the greeting message
+              waitMatch 5 conn $ \v ->
+                case Aeson.fromJSON v of
+                  Aeson.Success timedOutput ->
+                    guard $ output timedOutput == greeting
+                  _other -> Nothing
+
+              notHistoryMessage :: ServerOutput SimpleTx <- generate arbitrary
+              sendFromApiServer notHistoryMessage
+
+              -- Receive one more message. The messages we sent
+              -- before client connected are ignored as expected and client can
+              -- see only this last sent message.
+              received <- replicateM 1 (receiveData conn)
+
+              case traverse Aeson.eitherDecode received of
+                Left{} -> failure $ "Failed to decode messages:\n" <> show received
+                Right timedOutputs' -> do
+                  (output <$> timedOutputs') `shouldBe` [notHistoryMessage]
+              return ()
+
+  it "outputs tx as cbor or json depending on the client" $
+    monadicIO $ do
+      tx :: SimpleTx <- pick arbitrary
+      run $ do
+        withFreePort $ \port ->
+          withAPIServer @SimpleTx "127.0.0.1" (fromIntegral port) alice mockPersistence nullTracer noop $ \Server{sendOutput} -> do
+            let txValidMessage = TxValid{headId = HeadId "some-head-id", transaction = tx}
+                expectedTxCBORVal = Aeson.String . decodeUtf8 . Base16.encode $ serialize' tx
+                expectedTxJSONVal = toJSON tx
+                guardForValue v expected =
+                  case v of
+                    Aeson.Object km ->
+                      maybe
+                        Nothing
+                        (guard . (expected ==))
+                        (KeyMap.lookup "transaction" km)
```

Comment:
Could: simplify using fmap to `KeyMap.lookup "transaction" km <&> guard . (== expected)`

Reply:
It was `guard . (expected ==) =<< KeyMap.lookup "transaction" km`  in the end